components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java

```diff
@@ -1969,13 +2015,56 @@ public String addSubscription(APIIdentifier identifier, String userId, int appli
 
     public void removeSubscription(APIIdentifier identifier, String userId, int applicationId)
             throws APIManagementException {
-        apiMgtDAO.removeSubscription(identifier, applicationId);
+        API api = getAPI(identifier);
+        boolean isTenantFlowStarted = false;
+
+        if (tenantDomain != null && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+            PrivilegedCarbonContext.startTenantFlow();
+            isTenantFlowStarted = true;
```

Comment:
Why do we need to maintain a reference to check if the tenant flow has started within this particular logic? This is internally kept track of, so the application logic doesnt have to worry about it. If you try to close a tenant flow that is not started, then you'd either receive an IllegalStateException or something similar.